### PR TITLE
Issue #7609: Update doc for FinalClass #8996

### PR DESCRIPTION
**Pull request by ek08**
Issue #7609

Updated doc for FinalClass #8996

### CODE REVIEW

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java [hunk 1]
```diff
@@ -42,6 +42,45 @@
  * &lt;module name=&quot;FinalClass&quot;/&gt;
  * </pre>
  * <p>
+ * Example:
+ * </p>
+ * <pre>
+ * final class MyClass {
+ *   private MyClass(){ } //OK
```
Comment (Abhishek-kumar09) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java:49]: space after right parentheses.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java [hunk 2]
```diff
@@ -42,6 +42,45 @@
...
+ * }
+ *
+ * class MyClass {
+ *   private MyClass() { }
+ * } //violation,class should be declared final
```
Comment (Abhishek-kumar09) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java:54]: space after comma.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java [hunk 3]
```diff
@@ -42,6 +42,45 @@
...
+ *
+ * class MyClass {
+ *   int a;String str;
```
Comment (Abhishek-kumar09) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java:57]: put them on seperate line

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java (hunk 2 above)
Comment (Abhishek-kumar09) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java:54]: space after comma.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java (hunk 1 above)
Comment (Abhishek-kumar09) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java:49]: space after right parentheses.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java [hunk 4]
```diff
@@ -42,6 +42,45 @@
...
+ *   private MyClass(int a){
+ *   this.a=a;
+ *   str="dummy value";
```
Comment (Abhishek-kumar09) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java:60]: Put whitespaces there.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java [hunk 5]
```diff
@@ -42,6 +42,45 @@
...
+ *   }
+ *   public MyClass(String str){
+ *   this.str=str;
+ *   a=0;
+ *   } //OK,since it has a public constructor
```
Comment (Abhishek-kumar09) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java:65]: Same, use whitespaces to make your code clear to see.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java [hunk 6]
```diff
@@ -42,6 +42,45 @@
...
+ *   class MyClass{
+ *     private MyClass(){}
+ *   }
+ * } //OK,nested class in interface is always final
+ *
+ * public @interface Test {
+ *   public boolean enabled()
+ *   default true;
+ *   class MyClass{
+ *     private MyClass(){ }
```
Comment (Abhishek-kumar09) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java:79]: same

**Review by Abhishek-kumar09 — CHANGES_REQUESTED** [review]
Please have a look on the changes below.

**Comment by Abhishek-kumar09** [comment]
Please mention Issue # in your PR description, Also provide the screenshot of you changes by building site with
 `mvn clean site -Pno-validations`

**Comment by strkkk** [comment]
> Please mention Issue # in your PR description, Also provide the screenshot of you changes by building site with
mvn clean site -Pno-validations

@Abhishek-kumar09 There is no need to do it anymore, you can add comment "github, generate site" and github action job will create it and post link in PR

**Comment by ek08** [comment]
@Abhishek-kumar09 all the requested changes have been done.Please review it.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java [hunk 7]
```diff
@@ -42,6 +42,46 @@
...
+ * class MyClass {
+ *   private MyClass() { }
+ * } //violation, class should be declared final
+ *
+ * class MyClass {
+ *   int a;
+ *   String str;
+ *   private MyClass(int a) {
+ *   this.a=a;
+ *   str="dummy value";
```
Comment (Abhishek-kumar09) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java:61]:
```
this.a = a;
str = "dummy value";
```

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java [hunk 8]
```diff
@@ -42,6 +42,46 @@
...
+ *   }
+ *   public MyClass(String str) {
+ *   this.str=str;
+ *   a=0;
```
Comment (Abhishek-kumar09) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java:65]: same

**Review by Abhishek-kumar09 — CHANGES_REQUESTED** [review]

**Comment by ek08** [comment]
I have done these changes also.Please review it.

**Comment by romani** [comment]
@Abhishek-kumar09, please finish review and resolve all items is you are satisfied.

**Comment by romani** [comment]
@ek08 , please reply done on each review items, please do this always, it matter.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java (hunk 3 above)
Comment (ek08) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java (hunk 2 above)
Comment (ek08) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java]: Done

Comment (ek08) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java (hunk 1 above)
Comment (ek08) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java (hunk 4 above)
Comment (ek08) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java (hunk 5 above)
Comment (ek08) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java (hunk 6 above)
Comment (ek08) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java (hunk 7 above)
Comment (ek08) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java (hunk 8 above)
Comment (ek08) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java]: Done

**Comment by ek08** [comment]
> @ek08 , please reply "done" on each review items, please do this always, it matter.

I've replied done on all the reviews and will always do it from now on

**Comment by romani** [comment]
@ek08 , please squash all commits in one.

**Comment by romani** [comment]
Github, generate web site

**Comment by ek08** [comment]
> @ek08 , please squash all commits in one.

done

**Comment by romani** [comment]
In PR there should be 1 commit

**Comment by romani** [comment]
@ek08 , please share CLI output of check execution on example from xdoc, to make sure that line with violation match line where "violation" comment is present.

**Comment by ek08** [comment]
Prrof

**Comment by ek08** [comment]
I think I need to change the line where I have written violation it should be beside constructor.Am I right?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java (hunk 3 above)
Comment (Abhishek-kumar09) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java]: Resolved

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java (hunk 2 above)
Comment (Abhishek-kumar09) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java]: Resolved

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java (hunk 1 above)
Comment (Abhishek-kumar09) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java]: Resolved

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java (hunk 4 above)
Comment (Abhishek-kumar09) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java]: Resolved

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java (hunk 5 above)
Comment (Abhishek-kumar09) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java]: Resolved

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java (hunk 6 above)
Comment (Abhishek-kumar09) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java]: Resolved

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java (hunk 7 above)
Comment (Abhishek-kumar09) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java]: Resolved

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java (hunk 8 above)
Comment (Abhishek-kumar09) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java]: Resolved

**Comment by Abhishek-kumar09** [comment]
> I think I need to change the line where I have written violation it should be beside constructor.Am I right?

Yes, You have to check the cli for every code, and put `violations` or `OK` on the line which the cli is depicting.

**Comment by ek08** [comment]
I'm having problem in squashing.This is my final commit  https://github.com/checkstyle/checkstyle/pull/8996/commits/8faf0cdc301a106ad0487611d4a136ec7ed9e46a

**Comment by ek08** [comment]
> @ek08 , please squash all commits in one.

Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java [hunk 9]
```diff
@@ -42,6 +42,46 @@
  * &lt;module name=&quot;FinalClass&quot;/&gt;
  * </pre>
  * <p>
+ * Example:
+ * </p>
+ * <pre>
+ * final class MyClass {
+ *   private MyClass() { } //OK
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java:49]:
```
 * final class MyClass { // OK
 *   private MyClass() { }
```

comment "ok" should be on a line where "violation" can occur, please add space after "//".
Please apply both rules to all code in your "Example:"

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java [hunk 10]
```diff
@@ -42,6 +42,46 @@
...
+ * class MyClass { //violation, class should be declared final
+ *   private MyClass() { }
+ * }
+ *
+ * class MyClass {
+ *   int a;
+ *   String str;
+ *   private MyClass(int a) {
+ *   this.a = a;
+ *   str = "dummy value";
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java:61]: do it like
```
 *   int field1;
 *   String field2;
 *   private MyClass(int value) {
 *     this.field1 = value;
 *     this.field2 = "";
```

attention to indentation of code in example, and I removed extra code to let user focus on what matter - `ctors`.

**Review by romani — CHANGES_REQUESTED** [review]
please share CLI output in text form
see example - https://github.com/checkstyle/checkstyle/pull/7709

items:

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java (hunk 9 above)
Comment (ek08) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java (hunk 10 above)
Comment (ek08) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java]: Done

**Comment by ek08** [comment]
$ cat config.xml
&lt?xml version="1.0"?&gt
&lt!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd"  &gt

&ltmodule name="Checker"&gt
    &ltproperty name="severity" value="error"/&gt
    &ltmodule name="TreeWalker"&gt
		&ltmodule name="FinalClass"/&gt
    &lt/module&gt
&lt/module&gt

$ cat MyClass.java
final class MyClass { // OK
  private MyClass() { }
}

$ java -jar "-Duser.language=en -Duser.country=US" checkstyle-8.37-all.jar -c config.xml MyClass.java 
Starting audit...
Audit done.

$ cat MyClass.java
class MyClass { // violation, class should be declared final
  private MyClass() { }
}

$ java -jar "-Duser.language=en -Duser.country=US" checkstyle-8.37-all.jar -c config.xml MyClass.java 
Starting audit...
[ERROR] /home/ehtesham/MyClass.java:1:1: Class MyClass should be declared as final. [FinalClass]
Audit done.
Checkstyle ends with 1 errors.

$ cat MyClass.java
class MyClass {
  int field1;
  String field2;
  private MyClass(int value) {
    this.field1 = value;
    field2 = " ";
  }
  public MyClass(String value) { // OK, since it has a public constructor
    this.field2 = value;
    field1 = 0;
  }
}

$ java -jar "-Duser.language=en -Duser.country=US" checkstyle-8.37-all.jar -c config.xml MyClass.java 
Starting audit...
Audit done.

$ cat CheckInterface.java
interface CheckInterface
{
  class MyClass { // OK, nested class in interface is always final
    private MyClass() {}
  }
}

$ java -jar "-Duser.language=en -Duser.country=US" checkstyle-8.37-all.jar -c config.xml CheckInterface.java
Starting audit...
Audit done.

$ cat Test.java
public @interface Test {
  public boolean enabled()
  default true;
  class MyClass { // OK, class nested in an annotation is always final
    private MyClass() { }
  }
}

$ java -jar "-Duser.language=en -Duser.country=US" checkstyle-8.37-all.jar -c config.xml Test.java
Starting audit...
Audit done.

**Comment by Abhishek-kumar09** [comment]
> Indentation is proper but it is not getting previewed

Put you code between ```

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java [hunk 11]
```diff
@@ -42,6 +42,46 @@
...
+ * }
+ *
+ * class MyClass {
+ *   int field1;
+ *   String field2;
+ *   private MyClass(int value) {
+ *     this.field1 = value;
+ *     field2 = " ";
+ *   }
+ *   public MyClass(String value) { // OK, since it has a public constructor
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java:63]: Comment should be on line 56

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java [hunk 12]
```diff
@@ -42,6 +42,46 @@
...
+ *     this.field2 = value;
+ *     field1 = 0;
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java:65]: Please add `this` here and and in previous ctor

**Review by romani — CHANGES_REQUESTED** [review]
Items:

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java (hunk 11 above)
Comment (ek08) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java (hunk 12 above)
Comment (ek08) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java]: done

**Comment by romani** [comment]
Please always keep one commit, and resolve conflict

**Comment by romani** [comment]
Please do https://github.com/checkstyle/checkstyle/pull/8996#issuecomment-731292591

**Comment by ek08** [comment]
#8996

**Comment by ek08** [comment]
> Please do https://github.com/checkstyle/checkstyle/pull/8996#issuecomment-731292591

It means that I need to comment #8996

**Comment by romani** [comment]
You need to change your  comment to be readable, so you should use triple ` around code.

**Comment by ek08** [comment]
> You need to change your comment to be readable, so you should use triple ` around code.

done

**Comment by romani** [comment]
@ek08 , thanks a lot.

Please select one more issue from https://github.com/checkstyle/checkstyle/issues?q=is%3Aopen+is%3Aissue+label%3A%22gsoc+first+issue%22  and lets try to make it without to much to-from as it was in this PR.

**Review by romani — APPROVED** [review]

**Comment by ek08** [comment]
> @ek08 , thanks a lot.
> 
> Please select one more issue from https://github.com/checkstyle/checkstyle/issues?q=is%3Aopen+is%3Aissue+label%3A%22gsoc+first+issue%22 and lets try to make it without to much to-from as it was in this PR.

thanks a lot sir,will definitely try to solve other issues without this much to-from as in this one

**Pull request merged**